File: day02/src/main.rs
use advent_of_code_utils::read_file;

fn main() {
    let part1_solution = part1("input.txt");
    println!("Day 2 Part 1 Solution: {part1_solution}");
    
    let part2_solution = part2("input.txt");
    println!("Day 2 Part 2 Solution: {part2_solution}");
}

fn part1(path: &str) -> i64 {
    let lines = read_file(path);
    lines.iter()
        .map(|line| {
            line.split_whitespace()
                .map(|num| num.parse::<i64>().unwrap())
                .collect::<Vec<i64>>()
        })
        .filter(|nums| is_report_valid(nums)).count() as i64
    
}

fn part2(path: &str) -> i64 {
    let lines = read_file(path);
    lines.iter()
        .map(|line| {
            line.split_whitespace()
                .map(|num| num.parse::<i64>().unwrap())
                .collect::<Vec<i64>>()
        })
        .filter(|nums| {
            let valid_by_default = is_report_valid(nums);
            let valid_with_removal = nums.iter().enumerate().any(|(i, _)| {
                let mut nums_clone = nums.clone();
                nums_clone.remove(i);
                is_report_valid(&nums_clone)
            });
            valid_by_default || valid_with_removal       
        }).count() as i64
}

fn is_report_valid(nums: &Vec<i64>) -> bool {
    nums.windows(2).all(|w| {
        let abs_diff = (w[0] - w[1]).abs();
        abs_diff >= 1 && abs_diff <= 3
    }) && (nums.is_sorted() || nums.iter().rev().is_sorted())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        assert_eq!(part1("example.txt"), 2);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2("example.txt"), 4);
    }
}

File: day02/src/main.hs
import Data.List (tails, sort)

main :: IO()
main = do
    input <- readFile "input.txt"
    let reports = map (map read . words) (lines input) :: [[Int]]
    putStrLn $ "Solution for Day 02 -- Part 1: " ++ show (length $ filter part1ReportValid reports)
    putStrLn $ "Solution for Day 02 -- Part 2: " ++ show (length $ filter part2ReportValid reports)

windows :: Int -> [a] -> [[a]]
windows n = foldr (zipWith (:)) (repeat []) . take n . tails

isSorted :: [Int] -> Bool
isSorted [] = True
isSorted [_] = True
isSorted (x:y:xs) = x <= y && isSorted (y:xs)

reverseList :: [Int] -> [Int]
reverseList [] = []
reverseList (x:xs) = reverseList xs ++ [x]

removeIndex :: Int -> [a] -> [a]
removeIndex _ [] = []
removeIndex n (x:xs)
    | n < 0 = x:xs
    | n == 0 = xs
    | otherwise = x:removeIndex (n - 1) xs

part1ReportValid :: [Int] -> Bool
part1ReportValid report = all (\[x, y] -> abs (x - y) >= 1 && abs (x - y) <= 3 && (isSorted report || isSorted (reverseList report))) (windows 2 report)

part2ReportValid :: [Int] -> Bool
part2ReportValid report = do
    let validByDefault = part1ReportValid report
    let validByRemoval = any part1ReportValid [removeIndex i report | i <- [0..length report - 1]]
    validByDefault || validByRemoval

